Introduce reroute method on IngestDocument

felixbarny opened the pull request:
- Overrides _index
- Skips current pipeline
- Invokes default pipeline of new index

Requirement for
* https://github.com/elastic/elasticsearch/pull/76511

Combines https://github.com/elastic/elasticsearch/pull/85932 and https://github.com/elastic/elasticsearch/pull/85931

Fixes #83653

felixbarny committed:
Introduce redirect method on IngestDocument

- Overrides _index
- Skips current pipeline
- Invokes default pipeline of new index

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

felixbarny committed:
Add changelog

felixbarny committed:
Skipp full pipeline even if invoked via pipeline processor

dakrone commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -763,21 +751,78 @@ public void onFailure(Exception e) {
+        return new Pipelines(pipelineId, finalPipelineId);
+    }
+
+    private static class Pipelines implements Iterable<String> {
```
I think this is a reasonable abstraction, but I don't like that we make it mutable, and that we don't encapsulate enough here.

The `withoutDefaultPipeline` method makes me uncomfortable as its name makes it sound like it would return a new object rather than making a mutable change.

I also think we don't need to have `executePipelines(...)` take a boolean, we're treating this internally as though we'll always have a list of pipelines, but we could probably get away with passing a proper object instead `Iterator<String>`. I think it'd be much clearer that way.

felixbarny replied:
I abstracted the `Iterator<String>` and the boolean flag into a `PipelineIterator`, however it's not immutable. As that's the nature of iterators, I think that's fine.

The `executePipelines` needs to know about three properties of the current pipeline: the name (to add the name in the exception in case the pipeline itself can't be resolved), the pipeline itself, and whether the current pipeline is the final pipeline (if true, it's disallowed to override _index).

The `PipelineIterator` encapsulates that state and I think that makes it cleaner than before. Thanks for the suggestion 👍

dakrone commented on server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -903,6 +906,24 @@ public String toString() {
+        return skipCurrentPipeline;
+    }
+
+    void resetPipelineSkipping() {
+        skipCurrentPipeline = false;
+    }
```
Why doesn't this also reset `invokeDefaultPipelineOfDestination`? Also, it's unclear from the code why we need this method to be invoked.

felixbarny replied:
Good point, it needs to reset both. I've changed it to just be a single boolean.

dakrone commented on server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -903,6 +906,24 @@ public String toString() {
+        skipCurrentPipeline = true;
+    }
+
+    boolean isInvokeDefaultPipelineOfDestination() {
+        return invokeDefaultPipelineOfDestination;
+    }
+
+    boolean isSkipCurrentPipeline() {
+        return skipCurrentPipeline;
+    }
```
I don't think we need multiple boolean flags for this right? If we do, then it's not clear enough why they have to be separated (needs documentation)

felixbarny replied:
I've changed it to just be a single boolean and a single getter for it (`isRedirect()`). The downside is that `isRedirect()` is less expressive as `isSkipCurrentPipeline()` in the context of determining whether to skip the pipeline.
We could keep `isSkipCurrentPipeline()` and just return the `redirect` flag. But that feels a bit off as well.

felixbarny committed:
Encapsulate more state in PipelineIterator

felixbarny committed:
Only one boolean flag in IngestDocument

felixbarny committed:
Reset redirect at the end of the handler

felixbarny committed:
Apply spotless suggestions

felixbarny committed:
Rename method and add javadoc

felixbarny committed:
Reroute to remain

felixbarny committed:
Add test that final pipeline can't reroute

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ingest-document-redirect

joegallo commented:
I'm pretty sure this is going to fail the test introduced by https://github.com/elastic/elasticsearch/pull/94281, which means that this PR is a bugfix for https://github.com/elastic/elasticsearch/issues/83653. /cc @HiDAl

joegallo committed:
Update test

Since this PR fixes the linked bug

felixbarny committed:
Update docs/changelog/94000.yaml

elasticsearchmachine commented:
Hi @felixbarny, I've updated the changelog YAML for you.

breskeby commented:
@elasticsearchmachine test this please

joegallo committed:
Merge branch 'main' into ingest-document-redirect

joegallo committed:
Move this resetReroute call earlier

joegallo committed:
Pull this block out of the else

joegallo committed:
Reorder these blocks

If a final pipeline changes the indices such that a cycle is created,
it's more important to error that a final pipeline changed the indices
than that a cycle was created.

joegallo committed:
Add/tweak comments

joegallo committed:
Add more context to error message

HiDAl commented:
Yesterday I had to revert the test because it was failing due to warning headers. Today I pushed and merged PR #94388 which addresses the issue. Please rebase/merge on this PR @joegallo @felixbarny

felixbarny committed:
Merge remote-tracking branch 'origin/main' into ingest-document-redirect

felixbarny committed:
Adjust test to assert that final pipeline is not executed twice

felixbarny commented:
I've updated to main and re-adjusted the test. I ran that test locally and I can confirm that it prevents the final pipeline from running twice.

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ingest-document-redirect

joegallo committed:
Merge branch 'main' into ingest-document-redirect

joegallo committed:
Make PipelineIterator an iterator over a triple

joegallo committed:
Rename this method and add a docstring

joegallo committed:
The final pipeline slot should be isFinal, of course

joegallo committed:
Merge branch 'main' into ingest-document-redirect

joegallo committed:
Actually, getPipeline is all we need here

But make the null handling explicit rather than implicit

felixbarny commented:
> I'm not especially enamored with the way this tracks indexRecursionDetection -- there's a somewhat similar executedPipelines that's being tracked by IngestDocument itself. Having IngestDocument track the index changes and reroutes like that would simplify the parameters to executePipelines, and I think it would make later changes like adding the index history a bit simpler.

I completely agree with that. Ideally, ingest pipelines and scripts should not be able to modify the history so that we can be certain whether a document is sent directly to a target or whether it has been rerouted.

> I'll take on moving indexRecursionDetection over to IngestDocument as a refactoring in a separate PR.

Thanks!

> Let's merge this and start iterating on https://github.com/elastic/elasticsearch/pull/76511.

Thanks for the review and the approval! I'll wait for @dakrone's approval before merging.

dakrone reviewed (approved):
LGTM also, thanks for all the work on this Felix & Joe
